dhcpv6/option_vendor_opts.go

```diff
@@ -0,0 +1,161 @@
+	binary.BigEndian.PutUint32(buf[4:8], uint32(op.EnterpriseNumber))
+	for _, opt := range op.VendorOpts {
+		buf = append(buf, opt.ToBytes()...)
+	}
+	return buf
+}
+
+// Length returns the option length
+func (op *OptVendorOpts) Length() int {
+	l := 4 // 4 bytes for Enterprise Number
+	for _, opt := range op.VendorOpts {
+		l += 4 + opt.Length() // 4 bytes for Code and Length from Vendor
+	}
+	return l
+}
+
+// String returns a string representation of the VendorOpts data
+func (op *OptVendorOpts) String() string {
+	return fmt.Sprintf("OptVendorOpts{enterprisenum=%v, vendorOpts=%v}",
+		op.EnterpriseNumber, op.VendorOpts,
+	)
+}
+
+// ParseOptVendorOpts builds an OptVendorOpts structure from a sequence of bytes.
+// The input data does not include option code and length bytes.
+func ParseOptVendorOpts(data []byte) (*OptVendorOpts, error) {
+	opt := OptVendorOpts{}
+	if len(data) < 4 {
+		return nil, fmt.Errorf("Invalid vendor opts data length. Expected at least 4 bytes, got %v", len(data))
+	}
+	opt.EnterpriseNumber = binary.BigEndian.Uint32(data[:4])
+
+	var err error
+	opt.VendorOpts, err = VendorOptionsFromBytes(data[4:])
+	if err != nil {
+		return nil, err
+	}
+	return &opt, nil
+}
+
+// VendorOptionsFromBytes builds a slice of GenericOptions from a slice of bytes
+func VendorOptionsFromBytes(data []byte) ([]Option, error) {
+	// Parse a sequence of bytes until the end and build a list of options from
+	// it. Returns an error if any invalid option or length is found.
+	options := make([]Option, 0, 10)
+	if len(data) == 0 {
+		// no options, no party
+		return options, nil
+	}
+	if len(data) < 4 {
+		// cannot be shorter than option code (2 bytes) + length (2 bytes)
+		return nil, fmt.Errorf("Invalid options: shorter than 4 bytes")
+	}
+	idx := 0
+	for {
+		if idx == len(data) {
+			break
+		}
+		if idx > len(data) {
+			// this should never happen
+			return nil, fmt.Errorf("Error: reading past the end of options")
+		}
+		opt, err := VendParseOption(data[idx:])
+		if err != nil {
+			return nil, err
+		}
+		options = append(options, opt)
+		idx += opt.Length() + 4 // 4 bytes for type + length
+	}
+	return options, nil
+}
+
+// VendParseOption builds a GenericOption from a slice of bytes
+// We cannot use the exisitng ParseOption function in options.go because the
+// sub-options include codes specific to each vendor. There are overlaps in these
+// codes with RFC standard codes.
+func VendParseOption(dataStart []byte) (Option, error) {
+	// Parse a sequence of bytes as a single DHCPv6 option.
+	// Returns the option structure, or an error if any.
+	opt := &OptionGeneric{}
+
+	if len(dataStart) < 4 {
+		return opt, fmt.Errorf("Invalid DHCPv6 option: less than 4 bytes")
+	}
+	code := OptionCode(binary.BigEndian.Uint16(dataStart[:2]))
+	length := int(binary.BigEndian.Uint16(dataStart[2:4]))
+	if len(dataStart) < length+4 {
+		return opt, fmt.Errorf("Invalid option length for option %v. Declared %v, actual %v",
```

Comment:
I would specify this is a vendor option: "Invalid option length for vendor option"